storage/pebble/registers_pruner.go

```diff
@@ -0,0 +1,326 @@
+		}
+		defer it.Close()
+
+		var lastRegisterID flow.RegisterID
+		var keepKeyFound bool
+
+		for it.SeekGE(prefix); it.Valid(); it.Next() {
+			key := it.Key()
+
+			keyHeight, registerID, err := lookupKeyToRegisterID(key)
+			if err != nil {
+				return fmt.Errorf("malformed lookup key %v: %w", key, err)
+			}
+
+			if p.metrics != nil {
+				p.metrics.ElementVisited()
+			}
+
+			// New register prefix, reset the state
+			if !keepKeyFound || lastRegisterID != registerID {
+				keepKeyFound = false
+				lastRegisterID = registerID
+			}
+
+			if keyHeight <= pruneHeight {
+				if !keepKeyFound {
+					// Keep the first entry found for this registerID that is <= pruneHeight
+					keepKeyFound = true
+					continue
+				}
+
+				// Otherwise, mark this key for removal
+				// Create a copy of the key to avoid memory issues
+				keyCopy := make([]byte, len(key))
+				copy(keyCopy, key)
+				batchKeysToRemove = append(batchKeysToRemove, keyCopy)
+
+				if len(batchKeysToRemove) == itemsPerBatch {
+					// Perform batch delete
+					if err := p.batchDelete(batchKeysToRemove); err != nil {
+						return err
+					}
+					batchKeysToRemove = nil
+				}
+			}
+		}
+
+		if len(batchKeysToRemove) > 0 {
+			// Perform the final batch delete if there are any remaining keys
+			if err := p.batchDelete(batchKeysToRemove); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}(p.db)
+
+	if err != nil {
+		return err
+	}
+
+	duration := time.Since(start)
+
+	if p.metrics != nil {
+		p.metrics.Pruned(pruneHeight, duration)
+	}
+
+	return nil
+}
+
+// setupThrottleDelay sets up or reset a ticker for the throttle delay to manage system load
+// during pruning operations. This ensures that there is a small pause between pruning
+// each batch of registers.
+func (p *RegisterPruner) setupThrottleDelay() {
+	if p.throttleTicker == nil {
+		p.throttleTicker = time.NewTicker(p.pruneThrottleDelay)
+	} else {
+		p.throttleTicker.Reset(p.pruneThrottleDelay)
+	}
+}
+
+// batchDelete deletes the provided keys from the database in a single batch operation.
+// It creates a new batch, deletes each key from the batch, and commits the batch to ensure
+// that the deletions are applied atomically.
+//
+// Parameters:
+//   - lookupKeys: A slice of keys to be deleted from the database.
+//
+// No errors are expected during normal operations.
+func (p *RegisterPruner) batchDelete(lookupKeys [][]byte) error {
+	batch := p.db.NewBatch()
+	defer batch.Close()
+
+	for _, key := range lookupKeys {
+		keyHeight, registerID, _ := lookupKeyToRegisterID(key)
+		if err := batch.Delete(key, nil); err != nil {
+			return fmt.Errorf("failed to delete lookupKey: %w %d %v", err, keyHeight, registerID)
+		}
+	}
+
+	if err := batch.Commit(pebble.Sync); err != nil {
+		return fmt.Errorf("failed to commit batch: %w", err)
+	}
+
+	if p.metrics != nil {
+		p.metrics.NumberOfRowsPruned(uint64(len(lookupKeys)))
+	}
+
+	// Throttle to prevent excessive system load
+	<-p.throttleTicker.C
```

Comment:
rather than using a ticker, let's simplify to just using a sleep
```golang
select {
case <-ctx.Done():
case <-time.After(p.pruneThrottleDelay):
}
```

you can pass the context in from `loop()`

Reply:
[1ff10bc](https://github.com/onflow/flow-go/pull/6397/commits/1ff10bcc30ef9697bae0c939b4f6f347548bdc76)